crates/extricrate/src/lib.rs

```diff
@@ -25,32 +45,242 @@ pub mod dependencies {
+        #[error("file not readable")]
+        FileNotReadable,
+        #[error("path is not a crate")]
+        PathIsNotACrate,
+        #[error("linked module does not exists: {0}")]
+        ModuleDoesNotExists(String),
+    }
+
+    fn get_crate_entrypoint(crate_root: &Path) -> Option<PathBuf> {
+        // TODO: support multiple targets and custom paths different than src/main.rs or src/lib.rs
+
+        let cargo_toml = crate_root.join("Cargo.toml");
+        if !cargo_toml.exists() {
+            return None;
+        }
+
+        let main_rs = crate_root.join(Path::new("src/main.rs"));
+        if main_rs.exists() {
+            return Some(main_rs);
+        }
+
+        let lib_rs = crate_root.join(Path::new("src/lib.rs"));
+        if lib_rs.exists() {
+            return Some(lib_rs);
+        };
+        None
+    }
+
     /// List all the `use` statements in the crate, by file/module.
-    pub fn list_use_statements(crate_root: &std::path::Path) -> UseStatementMap {
-        todo!()
+    pub fn list_use_statements(
+        crate_root: &Path,
+    ) -> Result<UseStatementMap, ListUseStatementError> {
+        let mut files_visited = HashSet::new();
+        let mut files_to_visit = VecDeque::new();
+        let mut use_statement_map: UseStatementMap = HashMap::new();
+        let entry_point =
+            get_crate_entrypoint(crate_root).ok_or(ListUseStatementError::PathIsNotACrate)?;
+        let src_folder = entry_point
+            .parent()
+            .expect("Failed to get entry point parent folder");
+        files_to_visit.push_back((entry_point.clone(), "main".to_owned()));
+        while let Some(file_to_visit) = files_to_visit.pop_front() {
+            if files_visited.contains(&file_to_visit) {
+                continue;
+            }
+
+            if !file_to_visit.0.exists() {
+                return Err(ListUseStatementError::FileNotFound);
+            }
+
+            let content = read_to_string(&file_to_visit.0)
+                .map_err(|_| ListUseStatementError::FileNotReadable)?;
+
+            let parsed_file =
+                parse_file(&content).map_err(|_| ListUseStatementError::FileNotParsable)?;
+
+            let mut visitor = UseVisitor::new();
+            visitor.visit_file(&parsed_file);
+            for statement in visitor
+                .statements
+                .iter()
+                .flat_map(|dependency| &dependency.items)
+            {
+                match &statement.statement_type {
+                    UseStatementType::Simple(name) => {
+                        if is_local_import(&statement.module_name) {
+                            let file_to_visit = get_path_from_module_name(src_folder, name).ok_or(
+                                ListUseStatementError::ModuleDoesNotExists(name.to_string()),
+                            )?;
+                            dbg!(&files_to_visit);
+                            files_to_visit.push_back((file_to_visit, name.to_owned()));
```

Comment:
That's potentially an infinite loop if 2 modules reference each other, since you don't remember which one you already visited. But this will be fixed once we follow the files or the module declarations.

Reply:
Nevermind, you're right. Serves me right for reading code on my phone...